framework/src/materials/CoupledGradientMaterial.C

```diff
@@ -0,0 +1,50 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CoupledGradientMaterial.h"
+
+registerMooseObject("MooseApp", CoupledGradientMaterial);
+registerMooseObject("MooseApp", ADCoupledGradientMaterial);
+
+template <bool is_ad>
+InputParameters
+CoupledGradientMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Create a gradient material coupled to a variable.");
+  params.addRequiredParam<MaterialPropertyName>(
+      "grad_mat_prop",
+      "Name of gradient material property equal to the gradient of the coupled variable gradient "
+      "times the scalar.");
+  params.deprecateParam("grad_mat_prop", "gradient_material_name", "12/12/25");
+  params.addParam<MaterialPropertyName>(
+      "scalar_property",
+      1.0,
+      "Scalar material property multiplied by the coupled variable value and gradient.");
+  params.addRequiredCoupledVar("u", "The coupled variable");
+  params.deprecateCoupledVar("u", "coupled_variable", "12/12/25");
+  return params;
+}
+
+template <bool is_ad>
+CoupledGradientMaterialTempl<is_ad>::CoupledGradientMaterialTempl(
+    const InputParameters & parameters)
+  : Material(parameters),
+    _grad_mat_prop(declareGenericProperty<RealVectorValue, is_ad>("grad_mat_prop")),
+    _scalar_property(getGenericMaterialProperty<Real, is_ad>("scalar_property")),
+    _grad_u(coupledGenericGradient<is_ad>("u"))
+{
+}
```

Comment:
check if u is finite volume, and if so error because grad_u = 0 is not interesting

Reply:
hmm...whats the best way to do that?

Reply:
get a pointer to the variable and do:

if (var.isFV())
  mooseError

Reply:
Wait, is that true? This was originally used in `fvkernels/mms/grad-reconstruction.mat-rz` for a FV variable. If I replace the gradient with all zeros, it diffs.

Reply:
interesting! i didnt know this worked